Spigot-API-Patches/0058-Profile-Lookup-Events.patch

```diff
@@ -0,0 +1,221 @@
++
++    /**
++     * The UUID of the profile if it has already been provided by a plugin
++     * @return
++     */
++    @Nullable
++    public UUID getUUID() {
++        return uuid;
++    }
++
++    /**
++     * Sets the UUID for this player name. This will skip the initial API call to find the players UUID.
++     *
++     * However, if Profile Properties are needed by the server, you must also set them or else an API call will still be made.
++     * @param uuid
++     */
++    public void setUUID(@Nullable UUID uuid) {
++        this.uuid = uuid;
++    }
++
++    /**
++     * Get the properties for this profile
++     * @return
++     */
++    @Nonnull
++    public Multimap<String, Property> getProperties() {
++        return properties;
++    }
++
++    /**
++     * Completely replaces all Properties with the new provided properties
++     * @param properties
++     */
++    public void setProperties(Multimap<String, Property> properties) {
++        this.properties = new PropertyMap();
++        this.properties.putAll(properties);
++    }
++
++    /**
++     * Adds additional properties, without removing the original properties
++     * @param properties
++     */
++    public void addProperties(Multimap<String, Property> properties) {
++        this.properties.putAll(properties);
++    }
++
++    private static final HandlerList handlers = new HandlerList();
++
++    public HandlerList getHandlers() {
++        return handlers;
++    }
++
++    public static HandlerList getHandlerList() {
++        return handlers;
++    }
++
++    /**
++     * Wraps the Profile Repository so we can intercept all lookups
++     */
++    public static GameProfileRepository wrapProfileRepository(final GameProfileRepository orig) {
++        return (names, agent, callback) -> {
++            Set<String> unfoundNames = Sets.newHashSet();
++            for (String name : names) {
++                PreLookupProfileEvent event = new PreLookupProfileEvent(name);
++                event.callEvent();
++                if (event.getUUID() != null) {
++                    // Plugin provided UUI, we can skip network call.
++                    GameProfile gameprofile = new GameProfile(event.getUUID(), name);
++                    // We might even have properties!
++                    gameprofile.getProperties().putAll(event.getProperties());
++                    callback.onProfileLookupSucceeded(gameprofile);
++                } else {
++                    unfoundNames.add(name);
++                }
++            }
++
++            // Some things were not found.... Proceed to look up.
++            if (!unfoundNames.isEmpty() && orig != null) {
++                String[] namesArr = unfoundNames.toArray(new String[unfoundNames.size()]);
++                orig.findProfilesByNames(namesArr, agent, new PreProfileLookupCallback(callback));
++            }
++        };
++    }
++
++    private static class PreProfileLookupCallback implements ProfileLookupCallback {
```

Comment:
Goes with the above method, to make inspections happy to avoid having anon classes with 2 inner methods.